Crash when syncer is no longer able to connect to a physical cluster

**Describe the bug**
kcp crashes when it is unable to connect to a previously-registered physical cluster.

**To Reproduce**
Steps to reproduce the behavior:
1. `make build`
2. `./bin/kcp start --install-cluster-controller --install-namespace-scheduler --install-workspace-scheduler --auto-publish-apis --resources-to-sync="ingresses.networking.k8s.io,deployments.apps,services" --push-mode --token-auth-file contrib/demo/workspaceKubectlPlugin-script/kcp-tokens`
3. Start a physical cluster (e.g. kind)
4. Create a cluster.yaml for the physical cluster and apply it
5. Let the cluster be reconciled/synced at least once
6. Shut down the physical cluster
7. Wait for kcp to crash (it doesn't take all that long)

**Expected behavior**
No crash

**Additional context**
Here's the stack trace (testing from commit a7be6c6):

```
I0207 21:51:05.674008   82886 cluster.go:43] reconciling cluster "kind-us-east1"
I0207 21:51:05.680872   82886 syncer.go:165] Set up informer for /v1, Resource=namespaces
I0207 21:51:05.680899   82886 syncer.go:165] Set up informer for /v1, Resource=services
I0207 21:51:05.680911   82886 syncer.go:165] Set up informer for apps/v1, Resource=deployments
I0207 21:51:05.680922   82886 syncer.go:165] Set up informer for networking.k8s.io/v1, Resource=ingresses
E0207 21:51:15.412338   82886 memcache.go:179] couldn't get current server API group list: Get "https://127.0.0.1:53373/api?timeout=32s": read tcp 127.0.0.1:61746->127.0.0.1:53373: read: connection reset by peer
E0207 21:51:15.412364   82886 discovery.go:112] error mapping resource "ingresses.networking.k8s.io": Get "https://127.0.0.1:53373/api?timeout=32s": read tcp 127.0.0.1:61746->127.0.0.1:53373: read: connection reset by peer
E0207 21:51:15.731331   82886 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3192440, 0x50ce440})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:48 +0x75
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
I0207 21:51:15.744231   82886 panic.go:1038] Shutting down Cluster controller
E0207 21:51:15.744304   82886 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3192440, 0x50ce440})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001333220})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:48 +0x75
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2bde252]

goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001333220})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
 ```

--- a/cmd/syncer/main.go
+++ b/cmd/syncer/main.go
@@ -91,14 +91,11 @@ func main() {
 	defer cancel()
 
 	klog.Infoln("Starting workers")
-	syncer, err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), *clusterID, *fromCluster, numThreads)
-	if err != nil {
+	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), *clusterID, *fromCluster, numThreads); err != nil {
 		klog.Fatal(err)
 	}
 
 	<-ctx.Done()
 
 	klog.Infoln("Stopping workers")
-	syncer.Stop()
-	syncer.WaitUntilDone()
 }

--- a/pkg/reconciler/apiresource/controller.go
+++ b/pkg/reconciler/apiresource/controller.go
@@ -54,7 +54,7 @@ func NewController(
 	apiResourceImportInformer apiresourceinformer.APIResourceImportInformer,
 	crdInformer crdinfomer.CustomResourceDefinitionInformer,
 ) (*Controller, error) {
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-apiresource")
 
 	c := &Controller{
 		queue:                            queue,

--- a/pkg/reconciler/cluster/cluster.go
+++ b/pkg/reconciler/cluster/cluster.go
@@ -101,16 +101,35 @@ func (c *Controller) reconcile(ctx context.Context, cluster *clusterv1alpha1.Clu
 		}.String())
 	}
 
-	var upToDate bool
-	if c.syncerMode == SyncerModePull {
-		upToDate, err = isSyncerInstalledAndUpToDate(ctx, client, logicalCluster, c.syncerImage)
+	needsUpdate := false
+	switch c.syncerMode {
+	case SyncerModePull:
+		upToDate, err := isSyncerInstalledAndUpToDate(ctx, client, logicalCluster, c.syncerImage)
 		if err != nil {
 			klog.Errorf("error checking if syncer needs to be installed: %v", err)
 			return err
 		}
+
+		if !upToDate && groupResources.Len() > 0 {
+			needsUpdate = true
+		}
+	case SyncerModePush:
+		_, running := c.syncerCancelFuncs[cluster.Name]
+		if !running || !sets.NewString(cluster.Status.SyncedResources...).Equal(groupResources) {
+			needsUpdate = true
+		}
 	}
 
-	if !sets.NewString(cluster.Status.SyncedResources...).Equal(groupResources) || (!upToDate && groupResources.Len() > 0) {
+	if klog.V(2).Enabled() {
+		klog.V(2).InfoS("Determining if we need to start or update a syncer",
+			"synced-resources", cluster.Status.SyncedResources,
+			"group-resources", groupResources,
+			"equal", sets.NewString(cluster.Status.SyncedResources...).Equal(groupResources),
+			"needs-update", needsUpdate,
+		)
+	}
+	if needsUpdate {
+		klog.V(2).Info("Need to create/update syncer")
 		kubeConfig := c.kubeconfig.DeepCopy()
 
 		switch c.syncerMode {

--- a/pkg/reconciler/cluster/cluster.go
+++ b/pkg/reconciler/cluster/cluster.go
@@ -129,17 +129,20 @@ func (c *Controller) reconcile(ctx context.Context, cluster *clusterv1alpha1.Clu
 				return nil // Don't retry.
 			}
 
-			newSyncer, err := syncer.StartSyncer(ctx, upstream, downstream, groupResources, cluster.Name, logicalCluster, numSyncerThreads)
-			if err != nil {
+			syncerCtx, syncerCancel := context.WithCancel(ctx)
+			if err := syncer.StartSyncer(syncerCtx, upstream, downstream, groupResources, cluster.Name, logicalCluster, numSyncerThreads); err != nil {
 				klog.Errorf("error starting syncer in push mode: %v", err)
 				conditions.MarkFalse(cluster, clusterv1alpha1.ClusterReadyCondition, clusterv1alpha1.ErrorStartingSyncerReason, conditionsv1alpha1.ConditionSeverityError, "Error starting syncer in push mode: %v", err.Error())
+
+				syncerCancel()
+
 				return err
 			}
 
-			oldSyncer := c.syncers[cluster.Name]
-			c.syncers[cluster.Name] = newSyncer
-			if oldSyncer != nil {
-				oldSyncer.Stop()
+			oldSyncerCancel := c.syncerCancelFuncs[cluster.Name]
+			c.syncerCancelFuncs[cluster.Name] = syncerCancel
+			if oldSyncerCancel != nil {
+				oldSyncerCancel()
 			}
 
 			klog.Infof("started push mode syncer for cluster %s in logical cluster %s!", cluster.Name, logicalCluster)
@@ -216,13 +219,13 @@ func (c *Controller) cleanup(ctx context.Context, deletedCluster *clusterv1alpha
 
 		uninstallSyncer(ctx, client)
 	case SyncerModePush:
-		s, ok := c.syncers[deletedCluster.Name]
+		syncerCancel, ok := c.syncerCancelFuncs[deletedCluster.Name]
 		if !ok {
 			klog.Errorf("could not find syncer for cluster %q", deletedCluster.Name)
 			return
 		}
 		klog.Infof("stopping syncer for cluster %q", deletedCluster.Name)
-		s.Stop()
-		delete(c.syncers, deletedCluster.Name)
+		syncerCancel()
+		delete(c.syncerCancelFuncs, deletedCluster.Name)
 	}
 }

--- a/pkg/reconciler/cluster/controller.go
+++ b/pkg/reconciler/cluster/controller.go
@@ -41,7 +41,6 @@ import (
 	apiresourceinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/apiresource/v1alpha1"
 	clusterinformer "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/cluster/v1alpha1"
 	"github.com/kcp-dev/kcp/pkg/reconciler/apiresource"
-	"github.com/kcp-dev/kcp/pkg/syncer"
 )
 
 type SyncerMode int
@@ -92,12 +91,12 @@ func NewController(
 			clusterInformer.Informer().HasSynced,
 			apiResourceImportInformer.Informer().HasSynced,
 		},
-		syncerImage:     syncerImage,
-		kubeconfig:      kubeconfig,
-		resourcesToSync: resourcesToSync,
-		syncerMode:      syncerMode,
-		syncers:         map[string]*syncer.Syncer{},
-		apiImporters:    map[string]*APIImporter{},
+		syncerImage:       syncerImage,
+		kubeconfig:        kubeconfig,
+		resourcesToSync:   resourcesToSync,
+		syncerMode:        syncerMode,
+		syncerCancelFuncs: map[string]func(){},
+		apiImporters:      map[string]*APIImporter{},
 	}
 
 	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
@@ -145,7 +144,7 @@ type Controller struct {
 	kubeconfig                   clientcmdapi.Config
 	resourcesToSync              []string
 	syncerMode                   SyncerMode
-	syncers                      map[string]*syncer.Syncer
+	syncerCancelFuncs            map[string]func()
 	apiImporters                 map[string]*APIImporter
 	genericControlPlaneResources []schema.GroupVersionResource
 }

--- a/pkg/reconciler/cluster/controller.go
+++ b/pkg/reconciler/cluster/controller.go
@@ -79,7 +79,7 @@ func NewController(
 	resourcesToSync []string,
 	syncerMode SyncerMode,
 ) (*Controller, error) {
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-cluster")
 
 	c := &Controller{
 		queue:                    queue,

--- a/pkg/reconciler/deployment/controller.go
+++ b/pkg/reconciler/deployment/controller.go
@@ -49,7 +49,7 @@ const controllerName = "deployment"
 func NewController(cfg *rest.Config) *Controller {
 	client := appsv1client.NewForConfigOrDie(cfg)
 	kubeClient := kubernetes.NewForConfigOrDie(cfg)
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-deployment")
 	stopCh := make(chan struct{}) // TODO: hook this up to SIGTERM/SIGINT
 
 	csif := externalversions.NewSharedInformerFactoryWithOptions(clusterclient.NewForConfigOrDie(cfg), resyncPeriod)

--- a/pkg/reconciler/namespace/controller.go
+++ b/pkg/reconciler/namespace/controller.go
@@ -65,10 +65,10 @@ func NewController(
 	pollInterval time.Duration,
 ) *Controller {
 
-	resourceQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
-	gvrQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
-	namespaceQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
-	clusterQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	resourceQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-resource")
+	gvrQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-gvr")
+	namespaceQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-namespace")
+	clusterQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-cluster")
 
 	c := &Controller{
 		resourceQueue:  resourceQueue,

--- a/pkg/reconciler/workspace/controller.go
+++ b/pkg/reconciler/workspace/controller.go
@@ -60,7 +60,7 @@ func NewController(
 	workspaceInformer tenancyinformer.WorkspaceInformer,
 	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
 ) (*Controller, error) {
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-workspace")
 
 	c := &Controller{
 		queue:                 queue,

--- a/pkg/reconciler/workspaceindex/controller.go
+++ b/pkg/reconciler/workspaceindex/controller.go
@@ -58,8 +58,8 @@ func NewController(
 	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
 	index Index,
 ) (*Controller, error) {
-	orgQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
-	workspaceQueue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	orgQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-workspaceindex-org")
+	workspaceQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-workspaceindex-workspace")
 
 	c := &Controller{
 		orgQueue:              orgQueue,

--- a/pkg/reconciler/workspaceshard/controller.go
+++ b/pkg/reconciler/workspaceshard/controller.go
@@ -58,7 +58,7 @@ func NewController(
 	secretInformer coreinformer.SecretInformer,
 	workspaceShardInformer tenancyinformer.WorkspaceShardInformer,
 ) (*Controller, error) {
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-workspaceshard")
 
 	c := &Controller{
 		queue:                 queue,

--- a/pkg/syncer/specsyncer.go
+++ b/pkg/syncer/specsyncer.go
@@ -79,7 +79,7 @@ func NewSpecSyncer(from, to *rest.Config, syncedResourceTypes []string, clusterI
 	}
 	fromClient := fromClients.Cluster(logicalClusterID)
 	toClient := dynamic.NewForConfigOrDie(to)
-	return New(fromDiscovery, fromClient, toClient, KcpToPhysicalCluster, syncedResourceTypes, clusterID)
+	return New(clusterID, logicalClusterID, fromDiscovery, fromClient, toClient, KcpToPhysicalCluster, syncedResourceTypes, clusterID)
 }
 
 func (c *Controller) deleteFromDownstream(ctx context.Context, gvr schema.GroupVersionResource, namespace, name string) error {

--- a/pkg/syncer/statussyncer.go
+++ b/pkg/syncer/statussyncer.go
@@ -60,7 +60,7 @@ func NewStatusSyncer(from, to *rest.Config, syncedResourceTypes []string, cluste
 		return nil, err
 	}
 	toClient := toClients.Cluster(logicalClusterID)
-	return New(discoveryClient, fromClient, toClient, PhysicalClusterToKcp, syncedResourceTypes, clusterID)
+	return New(clusterID, logicalClusterID, discoveryClient, fromClient, toClient, PhysicalClusterToKcp, syncedResourceTypes, clusterID)
 }
 
 func (c *Controller) updateStatusInUpstream(ctx context.Context, gvr schema.GroupVersionResource, namespace string, obj *unstructured.Unstructured) error {

--- a/pkg/syncer/syncer.go
+++ b/pkg/syncer/syncer.go
@@ -28,8 +28,9 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/dynamic/dynamicinformer"
@@ -40,9 +41,11 @@ import (
 	"k8s.io/klog/v2"
 )
 
-const resyncPeriod = 10 * time.Hour
-const SyncerNamespaceKey = "SYNCER_NAMESPACE"
-const syncerApplyManager = "syncer"
+const (
+	resyncPeriod       = 10 * time.Hour
+	SyncerNamespaceKey = "SYNCER_NAMESPACE"
+	syncerApplyManager = "syncer"
+)
 
 // Direction indicates which direction data is flowing for this particular syncer
 type Direction string
@@ -59,34 +62,19 @@ type Syncer struct {
 	Resources    sets.String
 }
 
-func (s *Syncer) Stop() {
-	s.specSyncer.Stop()
-	s.statusSyncer.Stop()
-}
-
-func (s *Syncer) WaitUntilDone() {
-	<-s.specSyncer.Done()
-	<-s.statusSyncer.Done()
-}
-
-func StartSyncer(ctx context.Context, upstream, downstream *rest.Config, resources sets.String, cluster, logicalCluster string, numSyncerThreads int) (*Syncer, error) {
+func StartSyncer(ctx context.Context, upstream, downstream *rest.Config, resources sets.String, cluster, logicalCluster string, numSyncerThreads int) error {
 	specSyncer, err := NewSpecSyncer(upstream, downstream, resources.List(), cluster, logicalCluster)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	statusSyncer, err := NewStatusSyncer(downstream, upstream, resources.List(), cluster, logicalCluster)
 	if err != nil {
-		specSyncer.Stop()
-		return nil, err
+		return err
 	}
-	specSyncer.Start(ctx, numSyncerThreads)
-	statusSyncer.Start(ctx, numSyncerThreads)
+	go specSyncer.Start(ctx, numSyncerThreads)
+	go statusSyncer.Start(ctx, numSyncerThreads)
 
-	return &Syncer{
-		specSyncer:   specSyncer,
-		statusSyncer: statusSyncer,
-		Resources:    resources,
-	}, nil
+	return nil
 }
 
 type UpsertFunc func(ctx context.Context, gvr schema.GroupVersionResource, namespace string, unstrob *unstructured.Unstructured) error
@@ -104,10 +92,6 @@ type Controller struct {
 	direction Direction
 
 	syncerNamespace string
-
-	// set on start
-	ctx      context.Context
-	cancelFn context.CancelFunc
 }
 
 // New returns a new syncer Controller syncing spec from "from" to "to".
@@ -249,71 +233,52 @@ type holder struct {
 }
 
 func (c *Controller) AddToQueue(gvr schema.GroupVersionResource, obj interface{}) {
-	c.queue.AddRateLimited(holder{gvr: gvr, obj: obj})
+	c.queue.Add(holder{gvr: gvr, obj: obj})
 }
 
 // Start starts N worker processes processing work items.
 func (c *Controller) Start(ctx context.Context, numThreads int) {
-	c.ctx, c.cancelFn = context.WithCancel(ctx)
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
 
-	c.fromInformers.Start(c.ctx.Done())
-	c.fromInformers.WaitForCacheSync(c.ctx.Done())
+	c.fromInformers.Start(ctx.Done())
+	c.fromInformers.WaitForCacheSync(ctx.Done())
 
 	for i := 0; i < numThreads; i++ {
-		go c.startWorker(c.ctx)
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
 	}
+
+	<-ctx.Done()
 }
 
 // startWorker processes work items until stopCh is closed.
 func (c *Controller) startWorker(ctx context.Context) {
 	for {
-		select {
-		case <-ctx.Done():
-			klog.Info("stopping syncer worker")
-			return
-		default:
-			c.processNextWorkItem(ctx)
-		}
+		c.processNextWorkItem(ctx)
 	}
 }
-
-// Stop stops the syncer.
-func (c *Controller) Stop() {
-	c.queue.ShutDown()
-	c.cancelFn()
-}
-
-// Done returns a channel that's closed when the syncer is stopped.
-func (c *Controller) Done() <-chan struct{} { return c.ctx.Done() }
 
 func (c *Controller) processNextWorkItem(ctx context.Context) bool {
 	// Wait until there is a new item in the working queue
-	i, quit := c.queue.Get()
+	key, quit := c.queue.Get()
 	if quit {
 		return false
 	}
-	h := i.(holder)
+	h := key.(holder)
 
 	// No matter what, tell the queue we're done with this key, to unblock
 	// other workers.
-	defer c.queue.Done(i)
+	defer c.queue.Done(key)
 
-	err := c.process(ctx, h.gvr, h.obj)
-	c.handleErr(err, i)
-	return true
-}
-
-func (c *Controller) handleErr(err error, i interface{}) {
-	// Reconcile worked, nothing else to do for this workqueue item.
-	if err == nil {
-		c.queue.Forget(i)
-		return
+	if err := c.process(ctx, h.gvr, h.obj); err != nil {
+		// TODO(ncdc): see if we can plumb through a name to indicate which syncer this is (direction, clusters involved)
+		runtime.HandleError(fmt.Errorf("syncer failed to sync %q, err: %w", key, err))
+		c.queue.AddRateLimited(key)
 	}
 
-	// Give up and report error elsewhere.
-	c.queue.Forget(i)
-	utilruntime.HandleError(err)
-	klog.Errorf("Dropping key %q after failed retries: %v", i, err)
+	c.queue.Forget(key)
+
+	return true
 }
 
 // NamespaceLocator stores a logical cluster and namespace and is used

--- a/pkg/syncer/syncer.go
+++ b/pkg/syncer/syncer.go
@@ -82,6 +82,7 @@ type DeleteFunc func(ctx context.Context, gvr schema.GroupVersionResource, names
 type HandlersProvider func(c *Controller, gvr schema.GroupVersionResource) cache.ResourceEventHandlerFuncs
 
 type Controller struct {
+	name  string
 	queue workqueue.RateLimitingInterface
 
 	fromInformers dynamicinformer.DynamicSharedInformerFactory
@@ -95,10 +96,12 @@ type Controller struct {
 }
 
 // New returns a new syncer Controller syncing spec from "from" to "to".
-func New(fromDiscovery discovery.DiscoveryInterface, fromClient, toClient dynamic.Interface, direction Direction, syncedResourceTypes []string, clusterID string) (*Controller, error) {
-	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+func New(cluster, logicalCluster string, fromDiscovery discovery.DiscoveryInterface, fromClient, toClient dynamic.Interface, direction Direction, syncedResourceTypes []string, clusterID string) (*Controller, error) {
+	controllerName := string(direction) + "-" + logicalCluster + "-" + cluster
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-"+controllerName)
 
 	c := Controller{
+		name:            controllerName,
 		queue:           queue,
 		toClient:        toClient,
 		direction:       direction,
@@ -244,6 +247,8 @@ func (c *Controller) Start(ctx context.Context, numThreads int) {
 	c.fromInformers.Start(ctx.Done())
 	c.fromInformers.WaitForCacheSync(ctx.Done())
 
+	klog.InfoS("Starting syncer workers", "controller", c.name)
+	defer klog.InfoS("Stopping syncer workers", "controller", c.name)
 	for i := 0; i < numThreads; i++ {
 		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
 	}
@@ -271,8 +276,7 @@ func (c *Controller) processNextWorkItem(ctx context.Context) bool {
 	defer c.queue.Done(key)
 
 	if err := c.process(ctx, h.gvr, h.obj); err != nil {
-		// TODO(ncdc): see if we can plumb through a name to indicate which syncer this is (direction, clusters involved)
-		runtime.HandleError(fmt.Errorf("syncer failed to sync %q, err: %w", key, err))
+		runtime.HandleError(fmt.Errorf("syncer %q failed to sync %q, err: %w", c.name, key, err))
 		c.queue.AddRateLimited(key)
 	}
 

--- a/test/e2e/framework/accessory.go
+++ b/test/e2e/framework/accessory.go
@@ -110,30 +110,28 @@ func Ready(ctx context.Context, t TestingTInterface, port string) bool {
 }
 
 func waitForEndpoint(ctx context.Context, t TestingTInterface, port, endpoint string) {
-	var lastMsg string
-	var succeeded bool
-	loadCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
-	wait.UntilWithContext(loadCtx, func(ctx context.Context) {
+	var lastError error
+	if err := wait.PollImmediateWithContext(ctx, 100*time.Millisecond, 30*time.Second, func(ctx context.Context) (bool, error) {
 		url := fmt.Sprintf("http://[::1]:%s%s", port, endpoint)
 		resp, err := http.Get(url)
-		if err == nil {
-			body, err := ioutil.ReadAll(resp.Body)
-			if err != nil {
-				lastMsg = fmt.Sprintf("error reading response from %s: %v", url, err)
-				return
-			}
-			if resp.StatusCode != 200 {
-				lastMsg = fmt.Sprintf("unready response from %s: %v", url, string(body))
-				return
-			}
-			t.Logf("success contacting %s", url)
-			cancel()
-			succeeded = true
-		} else {
-			lastMsg = fmt.Sprintf("error contacting %s: %v", url, err)
+		if err != nil {
+			lastError = fmt.Errorf("error contacting %s: %w", url, err)
+			return false, nil
+
 		}
-	}, 100*time.Millisecond)
-	if !succeeded {
-		t.Errorf(lastMsg)
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			lastError = fmt.Errorf("error reading response from %s: %w", url, err)
+			return false, nil
+		}
+		if resp.StatusCode != 200 {
+			lastError = fmt.Errorf("unready response from %s: %v", url, string(body))
+			return false, nil
+		}
+
+		t.Logf("success contacting %s", url)
+		return true, nil
+	}); err != nil && lastError != nil {
+		t.Error(lastError)
 	}
 }

--- a/test/e2e/framework/kcp.go
+++ b/test/e2e/framework/kcp.go
@@ -307,22 +307,19 @@ func (c *kcpServer) loadCfg() error {
 }
 
 func (c *kcpServer) waitForEndpoint(client *rest.RESTClient, endpoint string) {
-	var lastMsg string
-	var succeeded bool
-	loadCtx, cancel := context.WithTimeout(c.ctx, 1*time.Minute)
-	wait.UntilWithContext(loadCtx, func(ctx context.Context) {
+	var lastError error
+	if err := wait.PollImmediateWithContext(c.ctx, 100*time.Millisecond, time.Minute, func(ctx context.Context) (bool, error) {
 		req := rest.NewRequest(client).RequestURI(endpoint)
 		_, err := req.Do(ctx).Raw()
-		if err == nil {
-			c.t.Logf("success contacting %s", req.URL())
-			cancel()
-			succeeded = true
-		} else {
-			lastMsg = fmt.Sprintf("error contacting %s: %v", req.URL(), err)
+		if err != nil {
+			lastError = fmt.Errorf("error contacting %s: %w", req.URL(), err)
+			return false, nil
 		}
-	}, 100*time.Millisecond)
-	if !succeeded && c.ctx.Err() == nil {
-		c.t.Errorf(lastMsg)
+
+		c.t.Logf("success contacting %s", req.URL())
+		return true, nil
+	}); err != nil && lastError != nil {
+		c.t.Error(lastError)
 	}
 }
 

--- a/test/e2e/framework/kcp.go
+++ b/test/e2e/framework/kcp.go
@@ -20,6 +20,7 @@ import (
 	"bufio"
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"os"
@@ -334,7 +335,7 @@ func (c *kcpServer) monitorEndpoint(client *rest.RESTClient, endpoint string) {
 	}
 	wait.UntilWithContext(ctx, func(ctx context.Context) {
 		_, err := rest.NewRequest(client).RequestURI(endpoint).Do(ctx).Raw()
-		if c.ctx.Err() != nil {
+		if errors.Is(err, context.Canceled) || c.ctx.Err() != nil {
 			return
 		}
 		if err != nil {

--- a/test/e2e/framework/t.go
+++ b/test/e2e/framework/t.go
@@ -18,6 +18,7 @@ package framework
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"testing"
 	"time"
@@ -102,7 +103,9 @@ func (t *T) flush() {
 	for {
 		select {
 		case err := <-t.errors:
-			t.T.Error(err)
+			if !errors.Is(err, context.Canceled) {
+				t.T.Error(err)
+			}
 		default:
 			return
 		}

